pkg/ruler/ruler.go

```diff
@@ -750,52 +771,200 @@ func (r *Ruler) getLocalRules(userID string) ([]*GroupStateDesc, error) {
 	return groupDescs, nil
 }
 
-func (r *Ruler) getShardedRules(ctx context.Context, userID string) ([]*GroupStateDesc, error) {
-	ring := ring.ReadRing(r.ring)
+func (r *Ruler) getShardedRules(ctx context.Context, userID string, quorumType QuorumType) ([]*GroupStateDesc, error) {
+	infoLogger := level.Info(r.logger)
+	debugLogger := level.Debug(r.logger)
+
+	type instanceResult struct {
+		res      []*GroupStateDesc
+		err      error
+		instance *ring.InstanceDesc
+	}
+
+	type groupCounter struct {
+		group *GroupStateDesc
+		count int
+	}
+
+	rulerRing := ring.ReadRing(r.ring)
 
 	if shardSize := r.limits.RulerTenantShardSize(userID); shardSize > 0 && r.cfg.ShardingStrategy == util.ShardingStrategyShuffle {
-		ring = r.ring.ShuffleShard(userID, shardSize)
+		rulerRing = r.ring.ShuffleShard(userID, shardSize)
 	}
 
-	rulers, err := ring.GetReplicationSetForOperation(RingOp)
+	rulers, err := rulerRing.GetReplicationSetForOperation(RingOp)
 	if err != nil {
 		return nil, err
 	}
 
+	if rulerRing.ReplicationFactor() <= 2 {
+		// strong quorums are only supported for RF >= 3
+		quorumType = Weak
```

Comment:
Should we return error if quorumType is Strong but replication factor is smaller than 3?

Reply:
Personally, I don't think we should return an error for this user-supplied parameter value based on the current server configuration when a reasonable default behaviour (quorum=weak) exists.  Whether quorum=weak is a reasonable default is debatable, though.   You could argue we should be able to support quorum=strong for RF=2, with quorum=strong requiring that both replicas have the same result, and not tolerating any ruler outage.

For RF=1, quorum=weak is the same as quorum=strong.